File: backend/app/schemas/ticket.py
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.domain.enums import TicketCategory, TicketPriority, TicketStatus
from app.domain.workflow import next_statuses


class SortField(StrEnum):
    CREATED_AT = "created_at"
    PRIORITY = "priority"


class SortOrder(StrEnum):
    ASC = "asc"
    DESC = "desc"


class TicketCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    title: str = Field(min_length=3, max_length=120)
    description: str = Field(min_length=10, max_length=5000)
    category: TicketCategory
    priority: TicketPriority


class StatusUpdate(BaseModel):
    status: TicketStatus


class TicketListQuery(BaseModel):
    status: TicketStatus | None = None
    category: TicketCategory | None = None
    priority: TicketPriority | None = None
    sort_by: SortField = SortField.CREATED_AT
    order: SortOrder = SortOrder.DESC
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class StatusChangeRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_status: TicketStatus | None
    to_status: TicketStatus
    changed_at: datetime


class TicketSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    category: TicketCategory
    priority: TicketPriority
    status: TicketStatus
    created_at: datetime
    updated_at: datetime


class TicketDetail(TicketSummary):
    description: str
    history: list[StatusChangeRead]

    @computed_field
    @property
    def allowed_transitions(self) -> list[TicketStatus]:
        return next_statuses(self.status)


class TicketPage(BaseModel):
    items: list[TicketSummary]
    total: int
    page: int
    page_size: int


class ErrorResponse(BaseModel):
    detail: str

File: frontend/src/types/ticket.ts
export type TicketStatus = "open" | "in_progress" | "resolved" | "closed";
export type TicketPriority = "low" | "medium" | "high" | "urgent";
export type TicketCategory = "it" | "facilities" | "hr" | "finance" | "other";

export type SortField = "created_at" | "priority";
export type SortOrder = "asc" | "desc";

export interface StatusChange {
  id: number;
  from_status: TicketStatus | null;
  to_status: TicketStatus;
  changed_at: string;
}

export interface TicketSummary {
  id: number;
  title: string;
  category: TicketCategory;
  priority: TicketPriority;
  status: TicketStatus;
  created_at: string;
  updated_at: string;
}

export interface TicketDetail extends TicketSummary {
  description: string;
  history: StatusChange[];
  allowed_transitions: TicketStatus[];
}

export interface TicketPage {
  items: TicketSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface TicketCreatePayload {
  title: string;
  description: string;
  category: TicketCategory;
  priority: TicketPriority;
}

export interface TicketListParams {
  status?: TicketStatus;
  category?: TicketCategory;
  priority?: TicketPriority;
  sort_by: SortField;
  order: SortOrder;
  page: number;
  page_size: number;
}
